update/common.go

```diff
@@ -112,3 +127,133 @@ func getAllMiniBlocksWithDst(metaBlock *block.MetaBlock, destShardID uint32) []b
 
 	return mbHdrs
 }
+
+// CheckDuplicates checks if there are duplicated miniBlocks in the given map
+func CheckDuplicates(
+	shardIDs []uint32,
+	mapBodyHandler map[uint32]data.BodyHandler,
+	marshalizer marshal.Marshalizer,
+	hasher hashing.Hasher,
+) error {
+	miniBlocks := make([]*block.MiniBlock, 0)
+	for _, shardID := range shardIDs {
+		bodyHandler := mapBodyHandler[shardID]
+		blockBody, ok := bodyHandler.(*block.Body)
+		if !ok {
+			return process.ErrWrongTypeAssertion
+		}
+
+		miniBlocks = append(miniBlocks, blockBody.MiniBlocks...)
+	}
+
+	mapMiniBlocks := make(map[string]struct{})
+	for _, miniBlock := range miniBlocks {
+		mbHash, err := core.CalculateHash(marshalizer, hasher, miniBlock)
+		if err != nil {
+			return err
+		}
+
+		_, duplicatesFound := mapMiniBlocks[string(mbHash)]
+		if duplicatesFound {
+			return ErrDuplicatedMiniBlocksFound
+		}
+
+		mapMiniBlocks[string(mbHash)] = struct{}{}
+	}
+
+	return nil
+}
+
+// CreateBody will create a block body after hardfork import
+func CreateBody(
+	shardIDs []uint32,
+	mapBodies map[uint32]data.BodyHandler,
+	mapHardForkBlockProcessor map[uint32]HardForkBlockProcessor,
+) ([]*MbInfo, error) {
+	mapPostMbs := make(map[uint32][]*MbInfo)
+	for _, shardID := range shardIDs {
+		hardForkBlockProcessor, ok := mapHardForkBlockProcessor[shardID]
+		if !ok {
+			return nil, ErrNilHardForkBlockProcessor
+		}
+
+		bodyHandler, postMbs, err := hardForkBlockProcessor.CreateBody()
+		if err != nil {
+			return nil, err
+		}
+
+		body, ok := bodyHandler.(*block.Body)
+		if !ok {
+			return nil, process.ErrWrongTypeAssertion
+		}
+
+		log.Debug("CreateBody",
+			"miniBlocks in body", len(body.MiniBlocks),
+			"postMbs", len(postMbs),
+		)
+
+		mapPostMbs[shardID] = postMbs
+		mapBodies[shardID] = bodyHandler
+	}
+
+	return getLastPostMbs(shardIDs, mapPostMbs), nil
+}
+
+// CreatePostBodies will create all the post block bodies after hardfork import
+func CreatePostBodies(
+	shardIDs []uint32,
+	lastPostMbs []*MbInfo,
+	mapBodies map[uint32]data.BodyHandler,
```

Comment:
why can't we work directly with *Body pointers directly? The method expects *Body anyway...

Reply:
refactored